src/services/refactors/extractType.ts

```diff
@@ -0,0 +1,141 @@
+                doTypedefChange(changes, file, name, info.firstStatement, info.selection) :
+                doTypeAliasChange(changes, file, name, info.firstStatement, info.selection, info.typeParameters));
+
+            const renameFilename = file.fileName;
+            const renameLocation = getRenameLocation(edits, renameFilename, name, /*preferLastLocation*/ false);
+            return { edits, renameFilename, renameLocation };
+        }
+    });
+
+    interface TypeAliasInfo { isJS: false; selection: TypeNode; firstStatement: Statement; typeParameters: ReadonlyArray<string>; }
+    interface TypedefInfo { isJS: true; selection: TypeNode; firstStatement: Statement & HasJSDoc; }
+    type Info = TypeAliasInfo | TypedefInfo;
+    function getRangeToExtract(context: RefactorContext): Info | undefined {
+        const { file, startPosition } = context;
+        const isJS = isSourceFileJS(file);
+        const current = getTokenAtPosition(file, startPosition);
+        const range = createTextRangeFromSpan(getRefactorContextSpan(context));
+
+        const selection = findAncestor(current, (node => node.parent && rangeContainsSkipTrivia(range, node, file) && !rangeContainsSkipTrivia(range, node.parent, file)));
+        if (!selection || !isTypeNode(selection)) return undefined;
+
+        if (isJS) {
+            // typeparam tag is not supported yet
+            return {
+                isJS,
+                selection,
+                firstStatement: Debug.assertDefined((findAncestor(selection, isStatementButNotBlockAndHasJSDoc))),
+            };
+        }
+        else {
+            const firstStatement = Debug.assertDefined((findAncestor(selection, isStatementButNotBlock)));
+            const typeParameters = checkAndCollectionTypeArguments(context.program.getTypeChecker(), selection, firstStatement, file);
+            if (!typeParameters) return undefined;
+
+            return { isJS, selection, firstStatement, typeParameters };
+        }
+    }
+
+    function isStatementButNotBlock(n: Node): n is Statement {
+        return n && isStatement(n) && !isBlock(n);
+    }
+
+    function isStatementButNotBlockAndHasJSDoc(n: Node): n is (Statement & HasJSDoc) {
+        return isStatementButNotBlock && hasJSDocNodes(n);
+    }
+
+    function rangeContainsSkipTrivia(r1: TextRange, node: Node, file: SourceFile): boolean {
+        return rangeContainsStartEnd(r1, skipTrivia(file.text, node.pos), node.end);
+    }
+
+    function checkAndCollectionTypeArguments(checker: TypeChecker, selection: TypeNode, statement: Statement, file: SourceFile): string[] | undefined {
+        let hasError = false;
+        const result: string[] = [];
+        visitor(selection);
+        return hasError ? undefined : result;
+
+        function visitor(node: Node) {
+            if (isTypeReferenceNode(node)) {
+                if (isIdentifier(node.typeName)) {

```

Comment:
why is it not necessary to handle the case where `node.typeName` is not an identifier but is an entityName?

Reply:
Could TypeParameter be a `QualifiedName`?
In `checkAndCollectionTypeArguments` we only iter the TypeParameter whitch defined in the statement

Reply:
i think in the common cases where you have to collect the type, the type name is an identifier and not a qualified name. but i'm really not sure, so that's why i asked.